### PR TITLE
fix(fold): don't include whitespace end regions

### PR DESCRIPTION
**Pull request by lewis6991**
Some languages that are difficult to parse via Treesitter may
incorrectly include whitespace lines at the end of regions. This can
make the calculated folds sub-optimal.

To rectify, use a new `trim!` directive for languages with this issue and account
for it when calculating folds.

Also added folds for Make

---

Here's a demo with Neovim's Makefile:

(left: original, right: fixed with this PR)

Screenshot 2022-04-21 at 10 48 01

### CODE REVIEW

**Comment by lewis6991** [comment]
@clason

**Comment by clason** [comment]
Where is that white(space)list maintained?

Wouldn't it make more sense to use `#offset` predicates in `fold` or `local` queries (similarly to `injections`)?

**Comment by lewis6991** [comment]
Maybe, but the offset is dynamic. The region basically ends just before the next non-whitespace char. Can we handle that?

**Comment by clason** [comment]
Hmm, good question. A "trim empty lines" predicate may be useful in general, if feasible.

**Comment by clason** [comment]
In any case, you could add `latex` to the list (and probably `markdown` as well).

**Comment by clason** [comment]
A `#trim!` could be doable; look at https://github.com/neovim/neovim/blob/5c4ec254784e7e92f61b69114c6091a198fe600f/runtime/lua/vim/treesitter/query.lua#L307-L345

(Could also be added locally via `vim.treesitter.query.add_predicate()`.)

**Comment by lewis6991** [comment]
I've tested `markdown`, that clearly benefits from this change. I'll trust you that `latex` should be added.

**Comment by clason** [comment]
Apparently, `latex` has bigger issues with folding... Maybe leave it out for now.

**Comment by lewis6991** [comment]
That was harder than I thought it would be.

**Comment by clason** [comment]
> That was harder than I thought it would be.

sorry :(

**Comment by lewis6991** [comment]
Nah don't worry. It just shows how wild-west the directives stuff is currently. That needs standardizing somewhat.

**Comment by clason** [comment]
PR/tracking issue welcome :)

**Comment by lewis6991** [comment]
Yeah, It will be needed if we want to devolve languages to separate modules.

File: lua/nvim-treesitter/fold.lua
Comment (theHamsta) [inline lua/nvim-treesitter/fold.lua:44]: a bad thing about this `metadata.XXX` is that it is not specific to a capture (there could be multiple within one query). I had this doubt when we first added `metadata.priority` but then concluded that for priority we could also apply it to all the captures in query. For `range` I would  expect that it its unambiguous to what it refers to especially since `trim!` has the capture as an argument.

Apart from that this is a very important feature that could be used to replace `lua/nvim-treesitter/tsrange.lua` and by that also solve the performance problems of nvim-treesitter. I think `range` feature has potential to generalize (maybe even to core after it is battle tested in the queries here)

Comment (theHamsta) [reply lua/nvim-treesitter/fold.lua:44]: Did clason's comment much later. I think this approach could be used also for `offset!` to be able to use offset also for other query types.

Comment (theHamsta) [reply lua/nvim-treesitter/fold.lua:44]: Again as with offset: it would probably be more convenient if such queries would replace/modify the found nodes instead of setting metadata with the goal that there is only one source of truth for the node range

Comment (lewis6991) [reply lua/nvim-treesitter/fold.lua:44]: > ... and by that also solve the performance problems of nvim-treesitter.

Interesting. Why do you think this solves the perf issues?

> I think this approach could be used also for offset! to be able to use offset also for other query types.

I based this on `offset!` which simply populates `metadata.content`. I'm not sure what I've done different here.

> Again as with offset: it would probably be more convenient if such queries would replace/modify the found nodes instead of setting metadata with the goal that there is only one source of truth for the node range

Agreed, I simply took the route of the least resistance here.

So what's needed here? Is the change good as is (...for now), or do we need to approach this differently. I only know half of what I'm doing, so I don't doubt there might be some better or simpler ways of doing this.

Comment (theHamsta) [reply lua/nvim-treesitter/fold.lua:44]: We definitely need a solution that works for multiple captures in a query. Otherwise, `range` in meta data would probably work perfectly and could work the same way also for offset! (to use offset! outside of the injection world)

Regarding perf, there is nothing bad with lua/nvim-treesitter/tsrange.lua. the problem is that `make-range!` (the only directive that uses it) doesn't use the new metadata API but the an old abstraction of nvim-treesitter's query API which is aloe because it constructs a high number of tables.

**Comment by lewis6991** [comment]
I've made the relevant changes assuming I understood the review comments correctly.

The arguments to `trim!` is now a list of capture id's and `metadata` is populated for each capture id.

**Comment by theHamsta** [comment]
@lewis6991 looks nice! Let's see how https://github.com/neovim/neovim/pull/18276 goes. Would be good to stay aligned with upstream :+1:

File: lua/nvim-treesitter/query_predicates.lua
Comment (theHamsta) [inline lua/nvim-treesitter/query_predicates.lua:134]: maybe `trim-lines!` would be more explicit. We might add a `trim!` latter that works like `offset!` but trims certain characters like quotes or regexes

Comment (lewis6991) [reply lua/nvim-treesitter/query_predicates.lua:134]: Perhaps, though `trim` is (IMO) well understood to mean strip whitespace around a string.

Even though it only strips full lines at the moment, the intention is that this can be improved to work more generally on any region with preceding and trailing whitespace.

For trimming other characters, I would just stick with offset or implement are more generalized substitute directive.

Comment (theHamsta) [reply lua/nvim-treesitter/query_predicates.lua:134]: yeah, when it will trim all whitespaces `trim` would be a proper name

**Comment by theHamsta** [comment]
This is still blocked by the up the upstream PR

**Comment by lewis6991** [comment]
I think the upstream PR needs another review.

**Comment by lewis6991** [comment]
I think with https://github.com/neovim/neovim/pull/18276 merged, this should be good to go.

**Comment by vigoux** [comment]
It seems from here that we want to trim the empty lines in all cases when folding.

Don't you think that it would be better to actually do that in `fold.lua` instead of adding a new predicate ?

**Comment by lewis6991** [comment]
Maybe. I don't know if whitespace should be included in folding for some languages.

**Comment by clason** [comment]
What is the status here? I think a `trim` directive could be useful in core as well.

Regarding folds: I believe I've heard general dissatisfaction with the current implementation and would encourage a new implementation in core for 0.9 (as the natural followup to upstreamed highlighting).

**Comment by lewis6991** [comment]
The change is good, just needs a rebase.

I think this would be useful in core too, but we should at least get it merged here for a while first.

> Regarding folds: I believe I've heard general dissatisfaction with the current implementation and would encourage a new implementation in core for 0.9 (as the natural followup to upstreamed highlighting).

Really? I'm really happy with folds. Is there anything specific that needs addressing?

There was one issue I noticed in regards to how they are updated, however I have this snippet in my config which has been working solidly for me:

```lua
require'nvim-treesitter'.define_modules {
  fold = {
    attach = function()
      vim.opt.foldexpr = 'nvim_treesitter#foldexpr()'
      vim.opt.foldmethod = 'expr'
      vim.cmd.normal'zx' -- recompute folds
    end,
    detach = function() end,
  }
}
```

**Comment by clason** [comment]
> I think this would be useful in core too, but we should at least get it merged here for a while first.

Makes sense, in particular while core is in stabilization for 0.8.

> Really? I'm really happy with folds. Is there anything specific that needs addressing?

I don't know; my impression was also that the fold queries were pretty solid, but someone (you?) really didn't like the `foldexpr` implementation...

> There was one issue I noticed in regards to how they are updated, however I have this snippet in my config which has been working solidly for me:

Could be included out-of-the-box?

**Review by clason — APPROVED** [review]

**Comment by lewis6991** [comment]
> I don't know; my impression was also that the fold queries were pretty solid, but someone (you?) really didn't like the foldexpr implementation...

That's purely from an efficiency POV. To resolve that we need to add a proper fold API to core (see https://github.com/neovim/neovim/pull/19155 and https://github.com/neovim/neovim/issues/19226)

But as far as the accuracy of folds and how they generally behave, I think they are pretty good.

> Could be included out-of-the-box?

It's not super-refined and probably has a bunch of edge cases that need dealing with. The main concern is that folding is all window local whereas extmarks and the like are all buffer local. But I don't know, maybe it's good enough for now?

**Pull request merged**
